core/src/main/java/org/apache/iceberg/PartitionsTable.java

```diff
@@ -93,16 +93,36 @@ private static StaticDataTask.Row convertPartition(Partition partition) {
     return StaticDataTask.Row.of(partition.key, partition.recordCount, partition.fileCount, partition.specId);
   }
 
-  private static Iterable<Partition> partitions(StaticTableScan scan) {
+  private static Iterable<Partition> partitions(Table table, StaticTableScan scan) {
     CloseableIterable<FileScanTask> tasks = planFiles(scan);
-
-    PartitionMap partitions = new PartitionMap(scan.table().spec().partitionType());
+    Types.StructType partitionType = Partitioning.partitionType(table);
+    PartitionMap partitions = new PartitionMap(partitionType);
     for (FileScanTask task : tasks) {
-      partitions.get(task.file().partition()).update(task.file());
+      PartitionData original = (PartitionData) task.file().partition();
+      PartitionData normalized = normalizePartition(original, partitionType);
+      partitions.get(normalized).update(task.file());
     }
     return partitions.all();
   }
 
+  private static PartitionData normalizePartition(PartitionData partition, Types.StructType normalizedPartitionSchema) {
+    Map<Integer, Object> fieldIdToValues = Maps.newHashMap();
```

Comment:
One thing we may want to consider here is caching these normalization by spec rather than recomputing the mapping for every partition value. Thinking about this code being run on millions of files

Reply:
Discussed offline, added cache of positional mappings to the final partition type, for each spec-id